src/lib.rs

```diff
@@ -143,53 +100,178 @@ impl Default for Options {
     }
 }
 
-
 #[derive(Clone, Copy, PartialEq, Debug)]
 enum State {
     Empty,
     Document,
     Attributes,
 }
 
-struct DepthData {
-    range: Range<usize>,
+struct DepthData<'a> {
+    element_name: Option<&'a str>,
     has_children: bool,
 }
 
+// This wrapper writer is so that we can make sure formatted strings are properly escaped too,
+// as we don't have access to the formatting stuff without a fmt::Write implementation, so
+// we provide it by wrapping the writer given to us while escaping appropriately any string to
+// be written, depending on the type of node we're writing.
+struct FmtWriter<W: Write> {
```

Comment:
Is this wrapper necessary? I don't quite understand the reason behind it.

Reply:
Well unfortunately yes, because all the fmt::Arguments stuff is private, so the only ways to use format strings is by implementing this fmt::Write wrapper (since io::Write implementation like io::Cursor don't implement it), or to use `format!(…)` beforehand, but that means we'd make potentially big string copies around all the time, which would kill the efficiency of the crate.

Now, I do not like this wrapper too, but I see no other convenient way around (that does not involve `format!()` with string copies).